### ISSUE TITLE
Stronger support for transaction management

### ISSUE THREAD

**Issue by brafdlog** [prisma/prisma#5002]
# Links

This issue is a place to discuss and document use cases related to database transactions. It is being linked from [this docs page](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/transactions#future-transaction-support-in-prisma-client).

The following feature requests relate to issues discussed here:

- Transactional Batch API: https://github.com/prisma/prisma-client-js/issues/667
- Atomic Operations: https://github.com/prisma/prisma-client-js/issues/655
- Long running transactions: https://github.com/prisma/prisma/issues/1844

# Original Issue
Currently (to the best of my knowledge) there is no way to control the transactions of the different operations we make.
When working with a relational database, transactions are a basic feature that is widely used. Allowing running multiple operations in the same transaction is vital in many use cases and could prevent people from using prisma in cases where it is important.
Two options to support this:
1. A bulk operation api that allows sending multiple operations in one call and running all of them in the same transaction.
2. An alternative more intrusive but powerful option would be to expose the transaction concept to the api - allow starting a transaction explicitly, getting a transaction id, running multiple operations using that transaction id, and explicitly committing the transaction.

I think that since prisma does not support only databases that work with transactions, the second option would result in an api that is not as clean. But it is worth considering.

**Comment by hugbubby** [comment]
I have to believe the conversation here from Prisma devs is masking some internal underlying technical issue that they have yet to solve, and that their team doesn't believe what they're actually saying in this thread and on their blog.  Otherwise, your team has gone insane and decided to willingly make a product that, for all practical purposes, cannot be used to safely read and write from databases. The suggestion that your users ought to develop their own in-house application level locks for each potentially contested database object is _patently crazy_, and I wish you would instead tell us why you actually cannot give us a long running transaction API. It'd be funny to read the pretzel logic involved in defending this position if it weren't so infuriating.

**Comment by sorenbs** [comment]
@thekevinbrown,

> For us the use case is we need to be able to read data, make decisions based on that data, then decide what to write. This needs to happen in a particular isolation level. Consider something like a purchase:

If you can encode the "make decisions" part in SQL, then this is already possible with `prisma.executeRaw`. We want to enable this use case without having to write raw SQL by enabling you to reference values from previous statements in a transaction. It will be a long time before we get to this.

If the "make decisions" part has to involve application code or even external service calls, then you can implement this with Prisma today using the Optimistic Concurrency pattern. The general idea is that you prepare the write but make it conditional of the underlaying data not having changed. [Here is an example](https://github.com/prisma/prisma/issues/1844#issuecomment-830860673) of how to do that with `prisma.executeRaw`. We plan to make this possible without having to writ raw sql as well. This is likely to happen much sooner.

I understand that this particular approach does not work in all cases, but it does enable you to implement your particular use case with Prisma today. I hope this helps.

**Comment by thekevinbrown** [comment]
@sorenbs, would technically work in some cases, but genuinely won't work for us at scale.

We've sometimes got a bunch of completely different writes that may or may not need to happen depending on all sorts of values across different functions that should work regardless of if they're called in a transaction context or not. We don't want to bind our business logic to isolation levels if possible. Doing it the way you're suggesting, we'd have to really tightly couple all of that business logic in SQL instead of being able to just go:

```
begin transaction
call function
  fetch some stuff
  if (condition) write stuff
fetch summary
if (another condition) write stuff
call function with args
  write stuff
commit transaction
```

And importantly have those functions continue to work outside of a need for a transaction as well. So basically, yes, in theory, but it'd make it much harder to maintain the code. I'll grant that it'd perform better your way, but most of these areas are not performance bottlenecks anyway, and speed of developers maintaining the code is more important than shaving a few ms off of an invocation on a non-bottleneck for us.

**Comment by Sytten** [comment]
@hugbubby Some of what you say has a sense of truth to it. Most people in this thread don't bother understanding how prisma (or DBMS for that matter) really work before starting to use the technology. Prisma 2 inherits from prisma 1 the client/server architecture. This means every time you make a request it actually first serializes to a graphql-ish requests, sends it to a socket (unless using napi direct bindings), interprets it, fetch a connection from the pool and then send the sql request (inverse path for the response). Implementing long running transactions in that context is very hard to do reliably. Do you do independent requests? If so, how do you link them together? What if the JS side never sends the commit, will you keep a connection stuck forever? If you put a timeout, how long is it? If you keep the http connnection alive, how do you send new requests? 

@thekevinbrown I agree with you on code maintainability, I had to write plpgsql recently due to not having read, decide, write in prisma and it was painful. But I dont agree that long running transaction are not a bottleneck, they are because they keep a DB connection busy. Lets say your database has a 100 connection total (recommended for 4gb of RAM for postgres), if you have a simple api and a worker with 3 instances each you cant really give more than 10 connections per instance. If you use long running transactions with slow logic (ie making network calls) just a bit too much, it is very easy to exhaust your connection pool. This is mostly apparent in outages where the network calls could take several seconds causing a cascading effect on your whole service instead of just that one operation.

So in general prisma is right to push people to avoid long running transactions, but it should probably allow it too since not all databases are scriptable nor are people willing to write that logic in those sql like languages.

**Comment by sorenbs** [comment]
Thank you @thekevinbrown 

There is some discussion going on about what an api for fully flexible transactions in Prisma Client could look like over in this other issue: https://github.com/prisma/prisma/issues/1844#issuecomment-842158100

I'd love to get your thoughts captured the as well.

**Comment by thekevinbrown** [comment]
@Sytten I agree that it takes up a connection for longer, and that is a bottleneck. What I'm saying is: Of the tradeoffs that we have, we'd rather take up a connection for longer so our business logic is easier to maintain and is decoupled from how the database works than couple it and make it less maintainable to solve a performance problem which doesn't cause us any pain today.

We don't currently have problems exhausting the connection pool using these long running transactions with another ORM, which is what I meant when I said "not a bottleneck for us".

**Comment by korredor11** [comment]
Hi, just to show you guys some use cases that are a little edgy. Can be done by using the nested write but can cause issues with some stock calculation that need to be performed.

I have a use case for my customers where they need to have stock separated by storage and readily available.  A running average of stock price is also required.

So i have

Product(stock_price), stock (warehouse_id, amount, batch, stock_price). When an buy order is completed, i need to update the order, the product stock column and the stock table all at the same time. The problem is that without serializable isolation level i can end up with some nasty nasty problems.

This can all be done using nested writes, but isolation of operation is not enforced. So if two buy orders are processed proximate times, an they have some product in common, problems happen.

I was reading the blog post about transactions and "thinking in better design" but i cannot find something that will maintain this working. (Yeah, sure i can do all those data running_avg stocks and etc on a "Calculated" field, but this is not fast when you have too much products and orders).

If someone could point an error of my thinking, i am very happy to listen what i'm doing wrong and what could be improved.

Also i was thinking on creation of functions on database where those transactions would run, then call this from the client, but manage versioning of database code is really bad, if i would need to update 30 instances of the database (Clients databases are isolated) and some problems of procedures/functions versioning happen. man that's not fun.

**Comment by matthewmueller** [comment]
All of the features linked have been implemented. Closing!

**Comment by kasir-barati** [comment]
Any thought on doing this: Returning `$transaction` to the user and pass it to the API interface as a argument. Something like this:

```ts
const transactionObj = prisma.createTransaction();
const user = await prisma.user.create({
  data: {
    email: 'elsa@prisma.io',
    name: 'Elsa Prisma',
  },
  { transactions: transactionObj }
})
const temp = prisma.temp.update({ where: { id: 1}, data: {f1: 1},  { transactions: transactionObj }})
```

**Comment by kasir-barati** [comment]
Any thought on doing this @brafdlog: Returning `$transaction` to the user and pass it to the API interface as a argument. Something like this:

```ts
const transactionObj = prisma.createTransaction();
const user = await prisma.user.create({
  data: {
    email: 'elsa@prisma.io',
    name: 'Elsa Prisma',
  },
  { transactions: transactionObj }
})
const temp = prisma.temp.update({ where: { id: 1}, data: {f1: 1},  { transactions: transactionObj }})
```
